contrib/verifybinaries/verify.py

```diff
@@ -47,136 +189,498 @@ def parse_version_string(version_str):
+    except urllib.request.HTTPError as e:
+        log.warning(f"HTTP request to {url} failed (HTTPError): {e}")
+    except Exception as e:
+        log.warning(f"HTTP request to {url} failed ({e})")
+    return (False, [])
+
+
+def verify_with_gpg(
+    signature_filename,
+    output_filename: t.Optional[str] = None
+) -> t.Tuple[int, str]:
+    args = [
+        'gpg', '--yes', '--decrypt',
+        '--output', output_filename if output_filename else '', signature_filename]
+
+    env = dict(os.environ, LANGUAGE='en')
+    result = subprocess.run(args, stderr=subprocess.STDOUT, stdout=subprocess.PIPE, env=env)
+    log.debug(f'Result from GPG ({result.returncode}): {result.stdout}')
+    return result.returncode, result.stdout.decode().rstrip()
+
+
+def remove_files(filenames):
+    for filename in filenames:
+        os.remove(filename)
+
+
+class SigData:
+    """GPG signature data as parsed from GPG stdout."""
+    def __init__(self, key: str, name: str, trusted: bool, status: str):
+        self.key = key
+        self.name = name
+        self.trusted = trusted
+        self.status = status
+
+    def __repr__(self):
+        return (
+            "SigData(%r, %r, trusted=%s, status=%r)" %
+            (self.key, self.name, self.trusted, self.status))
+
+
+def parse_gpg_result(output: str) -> ([SigData], [SigData], [SigData]):
+    """Returns good, unknown, and bad signatures from GPG stdout."""
+    good_sigs = []
+    unknown_sigs = []
+    bad_sigs = []
+    total_resolved_sigs = 0
+    curr_key = None
+
+    # Ensure that all lines we match on include a prefix that prevents malicious input
+    # from fooling the parser.
+    def line_begins_with(patt: str, line: str) -> bool:
+        return re.match(r'^\s*(gpg:)?(\s+)' + patt, line)
+
+    for i, line in enumerate(output):
+        if line_begins_with(r"using RSA key (0x[0-9a-fA-F]{16}|[0-9a-fA-F]{40})$", line):
+            if curr_key:
+                raise RuntimeError(
+                    f"WARNING: encountered a new sig without resolving the last ({curr_key}) - "
+                    "this could mean we have encountered a bad signature! check GPG output!")
+            curr_key = line.split('key ')[-1].strip()
+            assert len(curr_key) == 40 or (len(curr_key) == 18 and curr_key.startswith('0x'))
+
+        if line_begins_with(r"Can't check signature: No public key$", line):
+            if not curr_key:
+                raise RuntimeError("failed to detect signature being resolved")
+            unknown_sigs.append(SigData(curr_key, '', False, ''))
+            curr_key = None
+
+        if line_begins_with(r'Good signature from (".+")(\s+)(\[.+\])$', line):
+            if not curr_key:
+                raise RuntimeError("failed to detect signature being resolved")
+            name, status = parse_gpg_from_line(line)
+
+            # It's safe to index output[i + 1] because if we saw a good sig, there should
```

Comment:
There may well be another line in the good case, however my output looks like this for many keys:

```
  gpg: Good signature from "Antoine Poinsot <darosior@protonmail.com>" [unknown]
  gpg:                 aka "darosior <darosior@protonmail.com>" [unknown]
  gpg:                 aka "Antoine Poinsot <antoine@revault.dev>" [unknown]
  gpg:                 aka "darosior <darosior@ln.dev>" [unknown]
  gpg: WARNING: This key is not certified with a trusted signature!
```

So it shows up as `trusted = True`, when in fact it should not be

Reply:
Perhaps keep indexing until we don't reach a line which starts with `aka`, and then read that? This does feel quite brittle though.